Quality control flags in YAML stubs and VarBC for AMSUA

ibanos90 opened the pull request:
### Description
This PR is a follow up of #89 in which the feature branch is renamed and changes are made accordingly following that PR's comments/suggestions.

Here, the variable `biasCorrection` gets only the values `True` or `False` indicating that bias correction is not applied for pre-generated periods (PANDACArchive) since the observations are already bias corrected, and that variational bias correction is applied otherwise. The `biasCorrection` settings is available in both `variational` and `hofx` appications, but is only implemented so far for `variational`.  Setting `biasCorrection` to True for `variational` also turns on variational bias correction coefficient updating.  That new functionality enables the writing of the satellite bias correction coefficients at each cycle and uses the coefficients calculated from the previous cycle as prior estimates.

In addition, the YAML stubs are split and each part is added to a corresponding folder (base, filters, bias, filtersWithBias) for variational and hofx applications. In this way, a great flexibility is allowed in the sense of constructing the full application YAML from a modular set of sub-yamls.

Additionally, `satwnd` observations are enabled for both `variational` and `hofx` applications.  `satwnd` observations will only be used when `observations.resource` is set to either of `GladeRDAOnline` or `NCEPFTPOnline`, such that observations are converted to IODA format from BUFR.

### Issue closed
Closes #43
Closes #82
Closes #86

### Tests completed
 - [x] 3dvar_OIE120km_WarmStart
 - [x] 3denvar_OIE120km_WarmStart
 - [x] 3dvar_OIE120km_ColdStart
 - [x] 3denvar_O30kmIE60km_WarmStart
 - [x] eda_OIE120km_WarmStart

jjguerrette commented on PrepJEDI.csh:
```diff
@@ -251,15 +270,16 @@ set found = 0
-  endif
-  # check that instrument string matches at least one non-broken observation file link
-  find ${InDBDir}/${instrument}_obs_*.h5 -mindepth 0 -maxdepth 0
+    # check that instrument string matches at least one non-broken observation file link
+    find ${InDBDir}/${instrument}_obs_*.h5 -mindepth 0 -maxdepth 0
```
This check for missing observation files (L281-290) can be moved before and outside of the `subdir` loop.

ibanos90 replied:
You are right. I moved it.

jjguerrette commented on config/ObsPlugs/variational/filtersBias/amsua_aqua.yaml:
```diff
@@ -0,0 +1,273 @@
+                            0.400, 0.500, 0.300, 0.350, 0.350,
+                            0.450, 1.000, 1.500, 2.500, 2.500]
+#  Gross check
+#  - filter: Background Check
+#    threshold: 3
```
Remove commented `Background Check` filter in `filterBias/amsua_*.yaml`.

ibanos90 replied:
Removed.

jjguerrette commented on config/ObsPlugs/variational/filtersBias/amsua_aqua.yaml:
```diff
@@ -0,0 +1,273 @@
+    horizontal_mesh: {{RADTHINDISTANCE}}
+  - filter: GOMsaver
+    filename: {{OutDBDir}}{{MemberDir}}/{{geoPrefix}}_amsua_aqua.nc4
+  - filter: Background Check
+    threshold: 3.0
+    <<: *multiIterationFilter
```
Do we still need the 3*SIGMA background check?  Or does the "Gross Check" above carry out the same procedure?  Just checking.  Answer based on @junmeiban's YAMLS obviously.  Same applies for all other `filterBase/amsua_*.yaml`'s.

`<<: *multiIterationFilter` might need to be be copied to the "Gross check" to ensure it applies to all iterations.

ibanos90 replied:
You are right. Following @junmeiban's YAMLS we don't need the 3*SIGMA background check. I removed it and added the `multiIterationFilter` to the "Gross check"

jjguerrette commented on scenarios/RealTime.yaml:
```diff
@@ -16,6 +16,7 @@ model:
 variational:
   DAType: 3dvar
   nInnerIterations: [60,]
+  satelliteBias: VarBC
```
👍

jjguerrette commented on config/variational.csh:
```diff
@@ -77,5 +79,8 @@ endif
+$setLocal satelliteBias
+$setLocal fixedCoeff
+$setLocal fixedTlapmeanCov
+$setLocal DirsYamlBase
+$setLocal DirsYamlBiasFilters
```
I think these can be controlled in `config/variational.csh` instead of via YAML since they depend entirely on the setting for `satelliteBias`.  I would replace these two lines with this:
```
## Directories for YAML stubs
if ($satelliteBias == None) then
  set AppYamlDirs = (base filtersBase)
else
  set AppYamlDirs = (base bias filtersBias)
endif
```

Then use `AppYamlDirs` in `PrepJEDI.csh` just as you do now.

ibanos90 replied:
Thanks for this suggestion. I used it as indicated.

jjguerrette commented on config/hofx.csh:
```diff
@@ -15,3 +15,5 @@ setenv setLocal "source $setConfig $baseConfig $scenarioConfig hofx"
 
 $setLocal observations
 $setLocal nObsIndent
+$setLocal DirsYamlBase
+$setLocal DirsYamlBiasFilters
```
Similar comment to `config/variational.csh`.  These two lines can be changed to
```
## Directories for YAML stubs
set AppYamlDirs = (base filtersBase)
```

Also, you will need to set `satelliteBias` to None in `config/hofx.csh` to ensure that the `variational` `satelliteBias` setting does not get used.  That could happen as a result of nested `config/*.csh` files being sourced.  That is why `source config/AppTypeTEMPLATE.csh` is the last environment definition in `PrepJEDI.csh`.

ibanos90 replied:
Changed as suggested.

jjguerrette commented on PrepJEDI.csh:
```diff
@@ -218,6 +218,25 @@ foreach instrument ($observations)
+# Satellite bias correction
+# =========================
+if ( $satelliteBias == None ) then
+  set AppYamlDirs = ($DirsYamlBase)
+  set satBiasDir = None
+else if ( $satelliteBias == VarBC ) then
+  set AppYamlDirs = ($DirsYamlBiasFilters)
```
See comments for `config/variational.csh` and `config/hofx.csh`.  `DirsYamlBase` and `DirsYamlBiasFilters` can be removed.

ibanos90 replied:
Removed.

jjguerrette reviewed (commented):
Wow @ibanos90!  You pulled this together in a very short time this week.  Thank you.  Mostly bookkeeping comments.  This looks great!

liujake commented:
> I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?

Yes. That is a good point.

ibanos90 commented on PrepJEDI.csh:
```diff
@@ -218,6 +218,25 @@ foreach instrument ($observations)
+  # next cycle after FirstCycleDate
+  set nextFirstDate = `$advanceCYMDH ${FirstCycleDate} +${self_WindowHR}`
+  if ( ${thisValidDate} == ${nextFirstDate} ) then
+    set satBiasDir = $fixedCoeff/2018
```
You are right. I activated the initialVARBC variable again to define that directory.

ibanos90 commented:
> I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?

Thanks @junmeiban for pointing this out. Could you please let me know the path for the latest `global_satinfo.txt` file?

junmeiban commented:
> > I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?
> 
> Thanks @junmeiban for pointing this out. Could you please let me know the path for the latest `global_satinfo.txt` file?

Hi @liujake, I do not know where you download the file `global_satinfo.txt`. Would you please show @ibanos90 where can she download the latest `global_satinfo.txt`. Thanks

liujake commented:
https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/gfs.v16.1.6/fix/fix_gsi/global_satinfo.txt

ibanos90 commented:
> https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/gfs.v16.1.6/fix/fix_gsi/global_satinfo.txt

Thanks @liujake.

ibanos90 commented:
> > I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?
> 
> Thanks @junmeiban for pointing this out. Could you please let me know the path for the latest `global_satinfo.txt` file?

Just one more thing. The channel usage should be adjusted in the `use_flag` of the Useflag check and the Inter-channel check of each satellite, right?

ibanos90 commented:
Hi @jjguerrette. I am running some tests after the last changes committed and keep getting the following error each time that `variational.csh` is sourced:
```
(ERROR): invalid value for satelliteBias
(ERROR): variational.satelliteBias = None
```

if I comment out `$setLocal satelliteBias` in `variational.csh` the error doesn't show up and the tasks run successfully.  Do you have any other suggestion?

jjguerrette commented on scenarios/base/workflow.yaml:
```diff
@@ -144,3 +144,7 @@ workflow:
   FCVFWindowHR: 6
+
+  # maximum consecutive cycle points to be active at any time
+  # for CriticalPathType == normal
```
`normal` --> `Normal`

jjguerrette commented:
> Hi @jjguerrette. I am running some tests after the last changes committed and keep getting the following error each time that `variational.csh` is sourced:
> 
> ```
> (ERROR): invalid value for satelliteBias
> (ERROR): variational.satelliteBias = None
> ```
> 
> if I comment out `$setLocal satelliteBias` in `variational.csh` the error doesn't show up and the tasks run successfully. Do you have any other suggestion?

In the version you pushed to github, you have what I would think is the correct implementation:
```
set satelliteBias = "`$getLocalOrNone satelliteBias`"
```

ibanos90 commented:
> 

Yes, but then in L83: `$setLocal satelliteBias` which is what is giving the error I think

jjguerrette commented:
> > 
> 
> Yes, but then in L83: `$setLocal satelliteBias` which is what is giving the error I think

Only one of the `set` or `get` commands need to be used on any one YAML parameter.  There should never be duplicates in a single script.  Also, we should never need duplicates ever.  If duplicate YAML reading is necessary, which they are not in your case, then we need to redesign. So all you need is
```
set satelliteBias = "`$getLocalOrNone satelliteBias`"
```

Maybe keep it with the other bias settings in `config/variational.csh` to keep the file organized.

ibanos90 commented:
> > > 
> > 
> > 
> > Yes, but then in L83: `$setLocal satelliteBias` which is what is giving the error I think
> 
> Only one of the `set` or `get` commands need to be used on any one YAML parameter. There should never be duplicates in a single script. Also, we should never need duplicates ever. If duplicate YAML reading is necessary, which they are not in your case, then we need to redesign. So all you need is
> 
> ```
> set satelliteBias = "`$getLocalOrNone satelliteBias`"
> ```
> 
> Maybe keep it with the other bias settings in `config/variational.csh` to keep the file organized.

Thanks @jjguerrette.

ibanos90 commented on config/ObsPlugs/variational/filtersBias/amsua_n18.yaml:
```diff
@@ -0,0 +1,269 @@
+            err0:  [ 2.500,  2.200,  2.000,  0.550,  0.300,
+                     0.230,  0.230,  0.250,  0.250,  0.350,
+                     0.400,  0.550,  0.800,  3.000,  3.500]
+            err1:  [20.000, 18.000, 12.000,  3.000,  0.300,
+                     0.230,  0.230,  0.250,  0.250,  0.350,
+                     0.400,  0.550,  0.800,  3.000, 18.000]
```
I am not sure if I understand your comment. Are the errors in this PR still different than those from @junmeiban?

jjguerrette replied:
I do not know for sure.  I was under the impression that `err0` and `err` are supposed to be identical.  @junmeiban, please comment.

ibanos90 replied:
I am following this code `/glade/scratch/jban/pandac/testpr_220126/DA/2018041500/3denvar.yaml`

jjguerrette replied:
OK, I see what I was missing now.  Channels 5-14 (the ones we have been assuming are not impacted by clouds) have identical `err0` and `err1` across all AMSUA instruments in that `3denvar.yaml` and in this PR.  That is sufficient to meet @liujake's requirements.  It is confusing to have channels 1-4 and 15 show cloud-dependent errors for some instruments and non-cloud-dependent errors for other instruments, but those channels are not assimilated.  So there is no impact on the result.

I hope that in a later PR these non-impacting YAML settings can be documented more clearly with in-line comments or else made to look more uniform across instruments.  It is not required for this PR.  Marking this comment resolved now.

ibanos90 replied:
Ok, thanks @jjguerrette.

junmeiban replied:
Oh, I see... For amsua_aqua and amsua_n15, I modified all channels to use constant values no matter the channel is used or not. For others, I just modified channel 5 to channel 9 which we used. For the scene-dependent values, I think it's okay, because we do not use these channels.

junmeiban replied:
Thanks @jjguerrette. I just saw your comments. You are right.

junmeiban commented:
> > > I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?
> > 
> > 
> > Thanks @junmeiban for pointing this out. Could you please let me know the path for the latest `global_satinfo.txt` file?
> 
> Just one more thing. The channel usage should be adjusted in the `use_flag` of the Useflag check and the Inter-channel check of each satellite. Is this correct @junmeiban @jjguerrette?

Yes.

ibanos90 commented:
> > > > I remember, when I began to run case 2020, @liujake sent me latest `global_satinfo.txt` to check which channel used for recent GSI channel selection. Considering this is a real time run, maybe @ibanos90 should also check this file for the latest channel selection. Right @liujake ?
> > > 
> > > 
> > > Thanks @junmeiban for pointing this out. Could you please let me know the path for the latest `global_satinfo.txt` file?
> > 
> > 
> > Just one more thing. The channel usage should be adjusted in the `use_flag` of the Useflag check and the Inter-channel check of each satellite. Is this correct @junmeiban @jjguerrette?
> 
> Yes.

Thanks for confirming that @junmeiban.

junmeiban commented:
Hi @ibanos90,
I am using your feature branch now. 
(1) I would like double check with you:
To turn on updated VarBC function in 3denvar, I only need add the following line in scenarios/3denvar_OIE120km_WarmStart.yaml
```
   MPASGridDescriptor: OIE120km
 variational:
   DAType: 3denvar
+  satelliteBias: VarBC
```
(2) How can I turn on qc filters for conventional obs?

(3) I got the following errors when I test 3denvar_OIE120km_WarmStart scenario with updated VarBC at first cycle. I am still checking the problem:
```
 CRTM_K_Matrix(FAILURE) : Input data check failed for profile #35
 CRTM_K_Matrix(FAILURE) : 1 profiles failed
 ufo_radiancecrtm_simobs(FAILURE) :  Error calling CRTM (setTraj) K-Matrix Model for amsua_aqua on rank           86
Abort from ufo_radiancecrtm_simobs
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 86 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
 CRTM_K_Matrix(FAILURE) : 2 profiles failed
```

Thanks,
Junmei

ibanos90 commented:
> Hi @ibanos90, I am using your feature branch now. (1) I would like double check with you: To turn on updated VarBC function in 3denvar, I only need add the following line in scenarios/3denvar_OIE120km_WarmStart.yaml
> 
> ```
>    MPASGridDescriptor: OIE120km
>  variational:
>    DAType: 3denvar
> +  satelliteBias: VarBC
> ```
> 
> (2) How can I turn on qc filters for conventional obs?
> 
> (3) I got the following errors when I test 3denvar_OIE120km_WarmStart scenario with updated VarBC at first cycle. I am still checking the problem:
> 
> ```
>  CRTM_K_Matrix(FAILURE) : Input data check failed for profile #35
>  CRTM_K_Matrix(FAILURE) : 1 profiles failed
>  ufo_radiancecrtm_simobs(FAILURE) :  Error calling CRTM (setTraj) K-Matrix Model for amsua_aqua on rank           86
> Abort from ufo_radiancecrtm_simobs
> --------------------------------------------------------------------------
> MPI_ABORT was invoked on rank 86 in communicator MPI_COMM_WORLD
> with errorcode 1.
> 
> NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
> You may or may not see output from other processes, depending on
> exactly when Open MPI kills them.
> --------------------------------------------------------------------------
>  CRTM_K_Matrix(FAILURE) : 2 profiles failed
> ```
> 
> Thanks, Junmei

Hi @junmeiban, 
Sorry for the late reply and thanks so much for pointing out a caveat in the PR code for conventional qc when doing `VarBC`. Answering your questions:
(1) I would like to call your attention to two things:
- if you use the observations source `PANDACArchive`, then you can use any of the scenarios with `WarmStart` initialization type such as  the `3denvar_OIE120km_WarmStart.yaml` scenario. But, as mentioned in @jjguerrette's comments (https://github.com/NCAR/MPAS-Workflow/pull/92#discussion_r842958345), we do not do `VarBC` in that case because the observations used are already bias corrected. This was also discussed in our previous PANDAC meeting, but @jjguerrette could comment if I am wrong.
- if you do not use the `PANDACArchive` data source, then you can use the `3dvar_OIE120km_ColdStart.yaml` scenario, which you can configure to use observations from RDA on Cheyenne or NCEP FTP files online.

Once you have the appropriate scenario and data source configured, yes, the way to turn on the `VarBC` is by adding `satelliteBias: VarBC` in the variational section of the yaml scenario.

(2) This is a very good question and made me realize that I was not using the filters for conventional observations. Thanks!
In the last commit, I added the base filters files for the conventional and the other sensor_satellite to the `filtersBias` folder. This way, although it is repeating some files from `filtersBase`, it allows us a simpler implementation. I also added empty files for the other observations in the `bias` folder. We can come back to this later if needed or I can try something different if you think this is not the best way @jjguerrette. 

(3) I would recommend you to try the changes in the last commit with the necessary modifications for your case and re-run the experiment to check if the error persists. Please let me know if it works.

Thanks so much,
Ivette

ibanos90 commented on PrepJEDI.csh:
```diff
@@ -270,13 +283,22 @@ foreach instrument ($observations)
-  else
-    echo "${instrument} data is selected, but missing; NOT adding ${instrument} to the YAML"
-  endif
+  foreach subdir (${AppYamlDirs})
```
This is a very good point. Thanks for this comment @jjguerrette.  
In order to make sure that `VarBC` is not used when `PANDACArchive` is the data source, a conditional statement was added to `PrepJEDI.csh` in the last commit. This way the data source dependency is satisfied. However, it is still not instrument dependent. I added the filters and empty bias files to the corresponding folders. I thought this could be a cleaner way to implement it, but please let me know if you think otherwise. I can try a different approach.

junmeiban commented:
Many thanks @ibanos90 's update and detailed information. I am re-running it.

jjguerrette commented on PrepJEDI.csh:
```diff
@@ -219,6 +219,32 @@ foreach instrument ($observations)
+# =========================
+# Satellite bias correction
+# =========================
+if ( ${observations__resource} == "PANDACArchive" ) then
+  if ( $satelliteBias == None ) then
+    set satBiasDir = None
+  else
+    echo "$0 (ERROR): $satelliteBias is not supported for ${observations__resource}; exiting with failure"
+    exit 1
+  endif
+else if ( ${observations__resource} =~ *"Online"* ) then
```
I have something in mind that would allow this `observations__resource` dependency to be moved to the YAML and will avoid the duplicated filter files for conventional obs.  Do you mind if I make a few changes to this branch, particularly in the YAML?  It's a fairly small change I have in mind, but would allow us to merge this branch and move on.

ibanos90 replied:
That sounds good. Thanks!

jjguerrette replied:
Working on it now.  Hopefully will have it ready by the end of the day.  If not, early tomorrow.

jjguerrette commented:
New changes should work, mostly YAML layout.  Now there is one setting for `variational` and `hofx`, each, called `biasCorrection`.  Set it to `True` or `False`.  If more instruments are added that allow bias correction, they can be added to the lists in `scenarios/base/observations.yaml` under `instrumentsAllowingBiasCorrection`.  I changed some folder names too.  The main design of the PR is not touched.

ibanos90 commented:
> New changes should work, mostly YAML layout. Now there is one setting for `variational` and `hofx`, each, called `biasCorrection`. Set it to `True` or `False`. If more instruments are added that allow bias correction, they can be added to the lists in `scenarios/base/observations.yaml` under `instrumentsAllowingBiasCorrection`. I changed some folder names too. The main design of the PR is not touched.

Thanks @jjguerrette.

ibanos90 commented:
> > > Hi @ibanos90, I am using your feature branch now. (1) I would like double check with you: To turn on updated VarBC function in 3denvar, I only need add the following line in scenarios/3denvar_OIE120km_WarmStart.yaml
> > > ```
> > >    MPASGridDescriptor: OIE120km
> > >  variational:
> > >    DAType: 3denvar
> > > +  satelliteBias: VarBC
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > (2) How can I turn on qc filters for conventional obs?
> > > (3) I got the following errors when I test 3denvar_OIE120km_WarmStart scenario with updated VarBC at first cycle. I am still checking the problem:
> > > ```
> > >  CRTM_K_Matrix(FAILURE) : Input data check failed for profile #35
> > >  CRTM_K_Matrix(FAILURE) : 1 profiles failed
> > >  ufo_radiancecrtm_simobs(FAILURE) :  Error calling CRTM (setTraj) K-Matrix Model for amsua_aqua on rank           86
> > > Abort from ufo_radiancecrtm_simobs
> > > --------------------------------------------------------------------------
> > > MPI_ABORT was invoked on rank 86 in communicator MPI_COMM_WORLD
> > > with errorcode 1.
> > > 
> > > NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
> > > You may or may not see output from other processes, depending on
> > > exactly when Open MPI kills them.
> > > --------------------------------------------------------------------------
> > >  CRTM_K_Matrix(FAILURE) : 2 profiles failed
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Thanks, Junmei
> > 
> > 
> > Hi @junmeiban, Sorry for the late reply and thanks so much for pointing out a caveat in the PR code for conventional qc when doing `VarBC`. Answering your questions: (1) I would like to call your attention to two things:
> > 
> > * if you use the observations source `PANDACArchive`, then you can use any of the scenarios with `WarmStart` initialization type such as  the `3denvar_OIE120km_WarmStart.yaml` scenario. But, as mentioned in @jjguerrette's comments ([Quality control flags in YAML stubs and VarBC for AMSUA #92 (comment)](https://github.com/NCAR/MPAS-Workflow/pull/92#discussion_r842958345)), we do not do `VarBC` in that case because the observations used are already bias corrected. This was also discussed in our previous PANDAC meeting, but @jjguerrette could comment if I am wrong.
> > * if you do not use the `PANDACArchive` data source, then you can use the `3dvar_OIE120km_ColdStart.yaml` scenario, which you can configure to use observations from RDA on Cheyenne or NCEP FTP files online.
> > 
> > Once you have the appropriate scenario and data source configured, yes, the way to turn on the `VarBC` is by adding `satelliteBias: VarBC` in the variational section of the yaml scenario.
> > (2) This is a very good question and made me realize that I was not using the filters for conventional observations. Thanks! In the last commit, I added the base filters files for the conventional and the other sensor_satellite to the `filtersBias` folder. This way, although it is repeating some files from `filtersBase`, it allows us a simpler implementation. I also added empty files for the other observations in the `bias` folder. We can come back to this later if needed or I can try something different if you think this is not the best way @jjguerrette.
> > (3) I would recommend you to try the changes in the last commit with the necessary modifications for your case and re-run the experiment to check if the error persists. Please let me know if it works.
> > Thanks so much, Ivette
> 
> I am using JJ's latest commits. Now I have no idea to run my current 120km, 3denvar experiment by using this feature branch: use raw data, qc+varbc.
> 
> 1.According to your information, I should use 3dvar_OIE120km_ColdStart.yaml, but I do not know how to switch from warm start to coldstart. 2. How can I turn on qc+varbc? (Once I use `3dvar_OIE120km_ColdStart.yaml`, it can turn on qc+varbc automaticly?)

Hi @junmeiban. I forgot to mention in my previous comment that the ColdStart initialization type and therefore the `3dvar_OIE120km_ColdStart.yaml` scenario currently works only for `3dvar` as the yaml scenario indicates. At this moment, the case for 120km, raw data, QC+VarBC with `3denvar` is not implemented yet in the workflow. Sorry for the misleading directions.

junmeiban commented:
> > > > > Hi @ibanos90, I am using your feature branch now. (1) I would like double check with you: To turn on updated VarBC function in 3denvar, I only need add the following line in scenarios/3denvar_OIE120km_WarmStart.yaml
> > > > > ```
> > > > >    MPASGridDescriptor: OIE120km
> > > > >  variational:
> > > > >    DAType: 3denvar
> > > > > +  satelliteBias: VarBC
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > (2) How can I turn on qc filters for conventional obs?
> > > > > (3) I got the following errors when I test 3denvar_OIE120km_WarmStart scenario with updated VarBC at first cycle. I am still checking the problem:
> > > > > ```
> > > > >  CRTM_K_Matrix(FAILURE) : Input data check failed for profile #35
> > > > >  CRTM_K_Matrix(FAILURE) : 1 profiles failed
> > > > >  ufo_radiancecrtm_simobs(FAILURE) :  Error calling CRTM (setTraj) K-Matrix Model for amsua_aqua on rank           86
> > > > > Abort from ufo_radiancecrtm_simobs
> > > > > --------------------------------------------------------------------------
> > > > > MPI_ABORT was invoked on rank 86 in communicator MPI_COMM_WORLD
> > > > > with errorcode 1.
> > > > > 
> > > > > NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
> > > > > You may or may not see output from other processes, depending on
> > > > > exactly when Open MPI kills them.
> > > > > --------------------------------------------------------------------------
> > > > >  CRTM_K_Matrix(FAILURE) : 2 profiles failed
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > Thanks, Junmei
> > > > 
> > > > 
> > > > Hi @junmeiban, Sorry for the late reply and thanks so much for pointing out a caveat in the PR code for conventional qc when doing `VarBC`. Answering your questions: (1) I would like to call your attention to two things:
> > > > 
> > > > * if you use the observations source `PANDACArchive`, then you can use any of the scenarios with `WarmStart` initialization type such as  the `3denvar_OIE120km_WarmStart.yaml` scenario. But, as mentioned in @jjguerrette's comments ([Quality control flags in YAML stubs and VarBC for AMSUA #92 (comment)](https://github.com/NCAR/MPAS-Workflow/pull/92#discussion_r842958345)), we do not do `VarBC` in that case because the observations used are already bias corrected. This was also discussed in our previous PANDAC meeting, but @jjguerrette could comment if I am wrong.
> > > > * if you do not use the `PANDACArchive` data source, then you can use the `3dvar_OIE120km_ColdStart.yaml` scenario, which you can configure to use observations from RDA on Cheyenne or NCEP FTP files online.
> > > > 
> > > > Once you have the appropriate scenario and data source configured, yes, the way to turn on the `VarBC` is by adding `satelliteBias: VarBC` in the variational section of the yaml scenario.
> > > > (2) This is a very good question and made me realize that I was not using the filters for conventional observations. Thanks! In the last commit, I added the base filters files for the conventional and the other sensor_satellite to the `filtersBias` folder. This way, although it is repeating some files from `filtersBase`, it allows us a simpler implementation. I also added empty files for the other observations in the `bias` folder. We can come back to this later if needed or I can try something different if you think this is not the best way @jjguerrette.
> > > > (3) I would recommend you to try the changes in the last commit with the necessary modifications for your case and re-run the experiment to check if the error persists. Please let me know if it works.
> > > > Thanks so much, Ivette
> > > 
> > > 
> > > I am using JJ's latest commits. Now I have no idea to run my current 120km, 3denvar experiment by using this feature branch: use raw data, qc+varbc.
> > > 1.According to your information, I should use 3dvar_OIE120km_ColdStart.yaml, but I do not know how to switch from warm start to coldstart. 2. How can I turn on qc+varbc? (Once I use `3dvar_OIE120km_ColdStart.yaml`, it can turn on qc+varbc automaticly?)
> > 
> > 
> > Hi @junmeiban. I forgot to mention in my previous comment that the ColdStart initialization type and therefore the `3dvar_OIE120km_ColdStart.yaml` scenario currently works only for `3dvar` as the yaml scenario indicates. At this moment, the case for 120km, raw data, QC+VarBC with `3denvar` is not implemented yet in the workflow. Sorry for the misleading directions.
> 
> Oh I see. It's totally fine. I can try to modify it to my experiment. Thank you and JJ so much for working on this big PR.

Thank you so much for your help @ibanos90 . Now I can run raw data+varbc+qc 3denvar experiment.

ibanos90 commented:
> > > > > > Hi @ibanos90, I am using your feature branch now. (1) I would like double check with you: To turn on updated VarBC function in 3denvar, I only need add the following line in scenarios/3denvar_OIE120km_WarmStart.yaml
> > > > > > ```
> > > > > >    MPASGridDescriptor: OIE120km
> > > > > >  variational:
> > > > > >    DAType: 3denvar
> > > > > > +  satelliteBias: VarBC
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > (2) How can I turn on qc filters for conventional obs?
> > > > > > (3) I got the following errors when I test 3denvar_OIE120km_WarmStart scenario with updated VarBC at first cycle. I am still checking the problem:
> > > > > > ```
> > > > > >  CRTM_K_Matrix(FAILURE) : Input data check failed for profile #35
> > > > > >  CRTM_K_Matrix(FAILURE) : 1 profiles failed
> > > > > >  ufo_radiancecrtm_simobs(FAILURE) :  Error calling CRTM (setTraj) K-Matrix Model for amsua_aqua on rank           86
> > > > > > Abort from ufo_radiancecrtm_simobs
> > > > > > --------------------------------------------------------------------------
> > > > > > MPI_ABORT was invoked on rank 86 in communicator MPI_COMM_WORLD
> > > > > > with errorcode 1.
> > > > > > 
> > > > > > NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
> > > > > > You may or may not see output from other processes, depending on
> > > > > > exactly when Open MPI kills them.
> > > > > > --------------------------------------------------------------------------
> > > > > >  CRTM_K_Matrix(FAILURE) : 2 profiles failed
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > Thanks, Junmei
> > > > > 
> > > > > 
> > > > > Hi @junmeiban, Sorry for the late reply and thanks so much for pointing out a caveat in the PR code for conventional qc when doing `VarBC`. Answering your questions: (1) I would like to call your attention to two things:
> > > > > 
> > > > > * if you use the observations source `PANDACArchive`, then you can use any of the scenarios with `WarmStart` initialization type such as  the `3denvar_OIE120km_WarmStart.yaml` scenario. But, as mentioned in @jjguerrette's comments ([Quality control flags in YAML stubs and VarBC for AMSUA #92 (comment)](https://github.com/NCAR/MPAS-Workflow/pull/92#discussion_r842958345)), we do not do `VarBC` in that case because the observations used are already bias corrected. This was also discussed in our previous PANDAC meeting, but @jjguerrette could comment if I am wrong.
> > > > > * if you do not use the `PANDACArchive` data source, then you can use the `3dvar_OIE120km_ColdStart.yaml` scenario, which you can configure to use observations from RDA on Cheyenne or NCEP FTP files online.
> > > > > 
> > > > > Once you have the appropriate scenario and data source configured, yes, the way to turn on the `VarBC` is by adding `satelliteBias: VarBC` in the variational section of the yaml scenario.
> > > > > (2) This is a very good question and made me realize that I was not using the filters for conventional observations. Thanks! In the last commit, I added the base filters files for the conventional and the other sensor_satellite to the `filtersBias` folder. This way, although it is repeating some files from `filtersBase`, it allows us a simpler implementation. I also added empty files for the other observations in the `bias` folder. We can come back to this later if needed or I can try something different if you think this is not the best way @jjguerrette.
> > > > > (3) I would recommend you to try the changes in the last commit with the necessary modifications for your case and re-run the experiment to check if the error persists. Please let me know if it works.
> > > > > Thanks so much, Ivette
> > > > 
> > > > 
> > > > I am using JJ's latest commits. Now I have no idea to run my current 120km, 3denvar experiment by using this feature branch: use raw data, qc+varbc.
> > > > 1.According to your information, I should use 3dvar_OIE120km_ColdStart.yaml, but I do not know how to switch from warm start to coldstart. 2. How can I turn on qc+varbc? (Once I use `3dvar_OIE120km_ColdStart.yaml`, it can turn on qc+varbc automaticly?)
> > > 
> > > 
> > > Hi @junmeiban. I forgot to mention in my previous comment that the ColdStart initialization type and therefore the `3dvar_OIE120km_ColdStart.yaml` scenario currently works only for `3dvar` as the yaml scenario indicates. At this moment, the case for 120km, raw data, QC+VarBC with `3denvar` is not implemented yet in the workflow. Sorry for the misleading directions.
> > 
> > 
> > Oh I see. It's totally fine. I can try to modify it to my experiment. Thank you and JJ so much for working on this big PR.
> 
> Thank you so much for your help @ibanos90 . Now I can run raw data+varbc+qc 3denvar experiment.

Great @junmeiban! I am glad it worked.

jjguerrette commented:
> the case for 120km, raw data, QC+VarBC with `3denvar`

To be clear, 3denvar should work for 2018 with the raw data and QC+VarBC, but maybe not for `ColdStart` as the `InitializationType`.  @junmeiban, I can add another scenario that aligns better with your experiments if you like.  It would be very close to the `3denvar_OIE120km_WarmStart` with the `observations` and `variational.biasCorrection` settings changed to look like `3dvar_OIE120km_ColdStart`.

jjguerrette commented:
@ibanos90, I updated the PR message to reflect the most recent changes.

jjguerrette commented:
@ibanos90, this PR has my approval too, but I will not give a formal approval due to my being a partial contributor.

Great work!

ibanos90 commented:
> @ibanos90, I updated the PR message to reflect the most recent changes.

Thanks @jjguerrette. I ran the tests again and all completed successfully.

ibanos90 commented:
> @ibanos90, this PR has my approval too, but I will not give a formal approval due to my being a partial contributor.
> 
> Great work!

Ok. Thanks again for your contributions @jjguerrette. It was very helpful.

liujake reviewed (approved):
Great job! @ibanos90 @jjguerrette . That is a lot of work with 140 files touched!

ibanos90 commented:
Thanks @junmeiban for noting that the channel usage for clear-sky amsua was incorrect. I corrected it following https://www.nco.ncep.noaa.gov/pmb/codes/nwprod/gfs.v16.1.6/fix/fix_gsi/global_satinfo.txt for channels 5-9 and to monitor the rest of the channels. The **differences** between this PR and @junmeiban's code are:

amsua_n18  --> channels 5, 8 and 9 are monitored
amsua_n19  --> channel 7 is monitored

liujake commented:
This can be merged whenever it is ready.

ibanos90 commented:
Thank you @jjguerrette, @junmeiban and @liujake for reviewing this PR!
